### PR TITLE
langchain: add RankLLM Reranker

### PR DESCRIPTION
**Pull request by 16BitNarwhal**
Integrate RankLLM reranker (https://github.com/castorini/rank_llm) into LangChain

An example notebook is given in `docs/docs/integrations/retrievers/rankllm-reranker.ipynb`

### CODE REVIEW

**Comment by ronakice** [comment]
Great work @16BitNarwhal on integrating this to Langchain. Hopefully this brings in more interest in high precision and effective listwise paradigm for reranking models like RankZephyr!

File: libs/langchain/langchain/retrievers/document_compressors/rankllm_rerank.py [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+                    from rank_llm.rerank.zephyr_reranker import ZephyrReranker
+
+                    values["client"] = ZephyrReranker()
+                elif model_enum == ModelType.GPT:
+                    from rank_llm.rerank.rank_gpt import SafeOpenai
+                    from rank_llm.rerank.reranker import Reranker
+
+                    from langchain.llms.openai import OpenAI
+
+                    llm = OpenAI(
```
Comment (hwchase17) [inline libs/langchain/langchain/retrievers/document_compressors/rankllm_rerank.py:74]: you are importing just to access the api key? lets not do that, surely there has to be a more direct way

File: libs/langchain/langchain/retrievers/document_compressors/rankllm_rerank.py [hunk 2]
```diff
@@ -0,0 +1,127 @@
+from __future__ import annotations
+
+from copy import deepcopy
+from enum import Enum
+from typing import TYPE_CHECKING, Any, Dict, Optional, Sequence
+
+from langchain_core.documents import Document
+from langchain_core.pydantic_v1 import Extra, Field, PrivateAttr, root_validator
+
+from langchain.callbacks.manager import Callbacks
```
Comment (hwchase17) [inline libs/langchain/langchain/retrievers/document_compressors/rankllm_rerank.py:10]: should probably import from core?

**Review by hwchase17 — COMMENTED** [review]
this should go in langchain-community, not langchain, as it is a third party integration

**Comment by 16BitNarwhal** [comment]
Hey @hwchase17, thanks for the feedback! Made the following changes:
- Moved rankllm to community along with the integration test and updated example notebook
- Imported from langchain_core instead of langchain for callbacks
- Directly get openai API key from env instead of importing openai

Also, I'm a bit unsure about something. I noticed that [flashrank](https://github.com/langchain-ai/langchain/blob/893f06b5de4dccfc0c388a5e0f99715be27789dd/libs/langchain/langchain/retrievers/document_compressors/flashrank_rerank.py#L24) doesn't seem to be located in community. I see they support models like RankZephyr, which was built and is maintained by RankLLM team. Would it also be more appropriate for FlashRank to be in community?

**Comment by hwchase17** [comment]
> Also, I'm a bit unsure about something. I noticed that [flashrank](https://github.com/langchain-ai/langchain/blob/893f06b5de4dccfc0c388a5e0f99715be27789dd/libs/langchain/langchain/retrievers/document_compressors/flashrank_rerank.py#L24) doesn't seem to be located in community. I see they support models like RankZephyr, which was built and is maintained by RankLLM team. Would it also be more appropriate for FlashRank to be in community?

Yes it should be, thanks for flagging. we will move

**Comment by 16BitNarwhal** [comment]
Sounds good. And please let me know if there's anything else I need to do to get this merged 😁

**Comment by hwchase17** [comment]
> Sounds good. And please let me know if there's anything else I need to do to get this merged 😁

some test/lint issues, i can try to resolve

**Review by hwchase17 — APPROVED** [review]

**Pull request merged**